### PR TITLE
feat(hesai): add support for xt16

### PR DESCRIPTION
**Pull request by jemmmel**
## PR Type

- New Feature

## Related Links

[PCAP data](https://drive.google.com/file/d/1uKecBbVFgmKDmJbymWjCmcUQagybVGK4/view?usp=drive_link)

## Description

Added support for XT16 LiDAR, based on existing XT32 
- Added new header files and config
- Modified existing code with XT16 references
- Added test bag file and .pcd

## Remarks

Related to https://github.com/tier4/nebula/issues/137

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: nebula_ros/config/lidar/hesai/PandarXT16.param.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/**:
+  ros__parameters:
+    host_ip: 255.255.255.255
+    sensor_ip: 192.168.1.201
```
Comment (mojomex) [inline nebula_ros/config/lidar/hesai/PandarXT16.param.yaml:4]:
```suggestion
    host_ip: 127.0.0.1
    sensor_ip: 127.0.0.1
```

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+struct TailXT16
+{
+  uint8_t reserved1[10];
+  uint8_t return_mode;
+  uint16_t motor_speed;
+  DateTime<1900> date_time;
+  uint32_t timestamp;
+  uint8_t factory_information;
+};
```
Comment (mojomex) [inline nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp:38]: THis is identical to the one from XT32, consider creating a common `TailXT16_32` or do a `using TailXT16 = TailXT32;` so we have less duplicate code.

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#pragma pack(pop)
+
+}  // namespace hesai_packet
+
+class PandarXT16 : public HesaiSensor<hesai_packet::PacketXT16>
+{
+public:
+  static constexpr float min_range = 0.05f;
+  static constexpr float max_range = 120.0f;
+  static constexpr size_t max_scan_buffer_points = 256000;
```
Comment (mojomex) [inline nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp:58]: I think this should be `128000` = `16 (channels) * 2 (returns) * 360 (FoV) / 0.009  (highest resolution, if 5 Hz scans are used)`
```suggestion
  static constexpr size_t max_scan_buffer_points = 128000;
```

File: nebula_common/include/nebula_common/hesai/hesai_common.hpp
Comment (mojomex) [inline nebula_common/include/nebula_common/hesai/hesai_common.hpp:1]: Please also add `SensorModel::HESAI_PANDARXT16` in lines L436 and L477.

**Review by mojomex — CHANGES_REQUESTED** [review]
# Evaluation

## Code Review

Everything looks really god, thank you! I left some small suggestions, XT16 was missing in two switch statements.

## Unit Tests

Decoder tests are well formed and passing :+1: 

## PCAP Replay

Using a currently closed-source tool, I instantiated a dummy XT16 using the information in the PCAP. Nebula connects without a problem and pointcloud and diagnostics output work without a problem (see screenshot)

It seems that, [just like XT32](https://github.com/tier4/nebula/pull/208#issuecomment-2446170580), tthe power diagnostics have garbage values in them. Since we don't have the newest TCP protocol specification, this is curently a won't fix in Nebula, so as far as your PR is concerned, it's ok to leave as-is :ok_man: 

image

## Documentation

Could you add XT16 to the [list of supported sensors](https://github.com/tier4/nebula/tree/develop/docs/supported_sensors.md#L20), and [add a parameter page](https://github.com/tier4/nebula/tree/develop/mkdocs.yml#L51) for the model?

My apologies, the `develop` branch was a bit behind `main`, so not all doc changes were reflected there. Could you rebase your PR onto the updated `develop`? :bow:

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp (hunk 2 above)
Comment (jemmmel) [reply nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp]: OK

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp (hunk 3 above)
Comment (jemmmel) [reply nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_xt16.hpp]: OK

File: nebula_common/include/nebula_common/hesai/hesai_common.hpp
Comment (jemmmel) [reply nebula_common/include/nebula_common/hesai/hesai_common.hpp:1]: OK

**Comment by jemmmel** [comment]
All review comments done, submitting for review

**Comment by mojomex** [comment]
Looks good, thanks!
Could you rebase onto `develop` once again? Since it was updated when your PR already existed, there is a lot of now-duplicated history in your branch :bow:

**Comment by jemmmel** [comment]
Yeah no worries, I think my method was bad. Do you know how to do this without pulling in the unneeded history? 
Previously I tried
1. Update develop branch with tier4 changes
2. Switch to pr branch and git rebase develop

**Comment by mojomex** [comment]
@jemmmel Rebasing something with merge commits inbetween is quite painful, so the easiest method is to

```shell
git fetch --all
git reset --soft upstream/develop
```

and then re-create the commits by hand. It doesn't have to be the exact same commit history, just roughly divided into the different steps (add parameter file, add to nebula_common etc, add tests).

The rebase you tried just now lost a few changes, e.g. in `hesai_common`.

**Comment by jemmmel** [comment]
Hey @mojomex, thanks for the tips. Since the changes were minor and I was making a mess of the rebase, I have done some dectructive pushing, sorry about that. I also had an issue with signing commits. Anyway I think I have fixed it now, sorry about that

**Review by mojomex — APPROVED** [review]
LGTM, will merge when CI passes :tada: 

Thanks for your contribution!

**Comment by jemmmel** [comment]
Added missing vendor file to fix the JSON schema check failing

**Comment by mojomex** [comment]
@jemmmel Could you do

```shell
diff nebula_ros/schema/PandarXT16.schema.json nebula_ros/schema/PandarXT32.schema.json
```

And migrate over the slight differences from XT32? Sensor model enum value and some references to other schemas changed

**Comment by jemmmel** [comment]
> @jemmmel Could you do
> 
> ```shell
> diff nebula_ros/schema/PandarXT16.schema.json nebula_ros/schema/PandarXT32.schema.json
> ```
> 
> And migrate over the slight differences from XT32? Sensor model enum value and some references to other schemas changed. Then schema check should pass 🙂

Done now, missed some changes from the rebase!

**Pull request merged**